### PR TITLE
Optimise relabeling by re-using memory

### PR DESCRIPTION
**Pull request by bboreham**
Relabeling can be a large portion of Prometheus memory allocation, and hence overall performance.
Especially relabeling in service discovery.

Since relabelling works by applying different rules to the same label set, we can re-use slices and greatly reduce memory allocation.  Also re-use the Builder itself since `NewBuilder()` makes some speculative allocations.

The benchmark is somewhat lengthy, due to including realistic data.

```
name                  old time/op    new time/op    delta
Relabel/example-4       5.32µs ± 4%    4.10µs ± 2%  -22.92%  (p=0.008 n=5+5)
Relabel/kubernetes-4    50.0µs ± 7%    33.6µs ± 2%  -32.93%  (p=0.008 n=5+5)

name                  old alloc/op   new alloc/op   delta
Relabel/example-4       1.83kB ± 0%    0.74kB ± 0%  -59.66%  (p=0.000 n=4+5)
Relabel/kubernetes-4    26.3kB ± 0%     3.4kB ± 0%  -87.07%  (p=0.008 n=5+5)

name                  old allocs/op  new allocs/op  delta
Relabel/example-4         33.0 ± 0%      23.0 ± 0%  -30.30%  (p=0.008 n=5+5)
Relabel/kubernetes-4       106 ± 0%        44 ± 0%  -58.49%  (p=0.008 n=5+5)
```

### CODE REVIEW

File: model/relabel/relabel.go [hunk 1]
```diff
@@ -192,24 +192,31 @@ func (re Regexp) String() string {
 // are applied in order of input.
 // If a label set is dropped, nil is returned.
 // May return the input labelSet modified.
```
Comment (rfratto) [inline model/relabel/relabel.go:194]: Is this comment ever true? I can't see where relabel.Process ever modifies the input label set.

Comment (bboreham) [reply model/relabel/relabel.go:194]: I'm sure you're right about the current code, but it seems a useful freedom to reserve so I wouldn't call the comment "stale".

**Comment by bboreham** [comment]
I pushed another commit which re-does "[re-use source values slice](https://github.com/prometheus/prometheus/pull/11147/commits/1652a4586b58ec30ad806347f910533907ae5225)" in a simpler way, and reduces allocations by 1 more.

File: model/relabel/relabel.go [hunk 2]
```diff
@@ -192,24 +192,29 @@ func (re Regexp) String() string {
...
-func Process(labels labels.Labels, cfgs ...*Config) labels.Labels {
+func Process(lbls labels.Labels, cfgs ...*Config) labels.Labels {
+	lb := labels.NewBuilder(nil)
```
Comment (damnever) [inline model/relabel/relabel.go:196]: I'd like another interface called `ProcessWithBuilder`.

Comment (bboreham) [reply model/relabel/relabel.go:196]: OK; is that relevant to this PR?
If not, it would belong in a separate PR with its own justification.

Comment (damnever) [reply model/relabel/relabel.go:196]: I was thinking about using a pool of Builders here.

Comment (bboreham) [reply model/relabel/relabel.go:196]: That could also be done inside `Process` without changing the api.

File: model/relabel/relabel.go [hunk 3]
```diff
@@ -192,24 +192,29 @@ func (re Regexp) String() string {
...
 }
 
-func relabel(lset labels.Labels, cfg *Config) labels.Labels {
-	values := make([]string, 0, len(cfg.SourceLabels))
+func relabel(lset labels.Labels, cfg *Config, lb *labels.Builder) labels.Labels {
+	var va [16]string
+	values := va[:0]
+	if len(cfg.SourceLabels) > cap(values) {
+		values = make([]string, 0, len(cfg.SourceLabels))
+	}
```
Comment (codesome) [inline model/relabel/relabel.go:211]: How is this better than the previous?\

**Review by codesome — APPROVED** [review]
LGTM with a comment

File: model/relabel/relabel.go (hunk 3 above)
Comment (codesome) [reply model/relabel/relabel.go:211]: Also, would re-using the `values` slice between the `relabel` calls make it better?

Comment (bboreham) [reply model/relabel/relabel.go:211]: For relabeling rules with up to 16 source labels (surely most real-life examples), it uses an array on the stack, thus performs no heap allocation.

I tried re-using the slice; see https://github.com/prometheus/prometheus/pull/11147/commits/135e5171a5d92dc7b2f5fbff6be133880dd1e949.  It is more complicated and does one allocation in every case, compared to zero in most cases.

**Pull request merged**